src/peft/tuners/lora/layer.py

```diff
@@ -250,13 +250,37 @@ def unmerge(self) -> None:
                 self.weight.data -= self.get_delta_weight(active_adapter)
 
     def get_delta_weight(self, adapter) -> torch.Tensor:
-        return (
-            transpose(
-                self.lora_B[adapter].weight @ self.lora_A[adapter].weight,
-                self.fan_in_fan_out,
-            )
-            * self.scaling[adapter]
-        )
+        """
+        Compute the delta weight for the given adapter. In case users wants to merge the adapter weights that are in
+        float16 while being on CPU, we need to cast the weights to float32, perform the merge and then cast back to
+        float16.
+
+        Args:
+            adapter (str):
+                The name of the adapter for which the delta weight should be computed.
+        """
+        device = self.lora_B[adapter].weight.device
+        dtype = self.lora_B[adapter].weight.dtype
+
+        cast_to_fp32 = device.type == "cpu" and dtype == torch.float16
+
+        weight_A = self.lora_A[adapter].weight
+        weight_B = self.lora_B[adapter].weight
+
+        if cast_to_fp32:
+            weight_A = weight_A.float()
+            weight_B = weight_B.float()
+
+        output_tensor = transpose(weight_B @ weight_A, self.fan_in_fan_out) * self.scaling[adapter]
```

Comment:
I wonder if we should put all the code starting from here until the `return` into a `try ... finally` to ensure that we always cast the weights back to their original dtype, even if there is an error. Might be overkill, not sure.

Too bad there isn't a PyTorch context manager that would allow something like this:

```python
with torch.force_dtype('float32'):
    output_tensor = transpose(weight_B @ weight_A, self.fan_in_fan_out) * self.scaling[adapter]
```

Would be quite handy here but I couldn't find anything that would do that.